File: packages/mirascript/src/vm/lib/mod/matrix/bin-ops.ts
import { Cp } from '../../../checkpoint.js';
import { $Add, $Call, $Div, $Mul, $Sub } from '../../../operations/index.js';
import { isVmConst, type VmConst } from '../../../types/index.js';
import { VmLib, expectCallable, expectConst, throwError } from '../../helpers.js';
import { num, size } from './helpers.js';
const { max } = Math;

/** 逐项操作 */
function entrywiseImpl(
    a: VmConst,
    b: VmConst,
    f: (a: VmConst, b: VmConst) => VmConst,
    vvf?: (va: readonly VmConst[], vb: readonly VmConst[], ar: number, br: number) => VmConst,
    mmf?: (
        ma: readonly VmConst[][],
        mb: readonly VmConst[][],
        ar: number,
        ac: number,
        br: number,
        bc: number,
    ) => VmConst,
    vmf?: (va: readonly VmConst[], mb: readonly VmConst[][], al: number, br: number, bc: number) => VmConst,
    mvf?: (ma: readonly VmConst[][], vb: readonly VmConst[], ar: number, ac: number, bl: number) => VmConst,
): VmConst {
    let [ar, ac] = size(a);
    let [br, bc] = size(b);

    if (ar == null) {
        if (br == null) {
            // s/s
            return f(a, b);
        } else if (bc == null) {
            // s/v
            const result: VmConst[] = [];
            for (let r = 0; r < br; r++) {
                const bItem = (b as VmConst[])[r] ?? null;
                result[r] = f(a, bItem);
            }
            return result;
        } else {
            // s/m
            const result: VmConst[][] = [];
            for (let r = 0; r < br; r++) {
                const bRow = (b as VmConst[][])[r] ?? [];
                const rRow: VmConst[] = [];
                result[r] = rRow;
                for (let c = 0; c < bc; c++) {
                    const bItem = bRow[c] ?? null;
                    rRow[c] = f(a, bItem);
                }
            }
            return result;
        }
    }
    if (br == null) {
        if (ac == null) {
            // v/s
            const result: VmConst[] = [];
            for (let r = 0; r < ar; r++) {
                const aItem = (a as VmConst[])[r] ?? null;
                result[r] = f(aItem, b);
            }
            return result;
        } else {
            // m/s
            const result: VmConst[][] = [];
            for (let r = 0; r < ar; r++) {
                const aRow = (a as VmConst[][])[r] ?? [];
                const rRow: VmConst[] = [];
                result[r] = rRow;
                for (let c = 0; c < ac; c++) {
                    const aItem = aRow[c] ?? null;
                    rRow[c] = f(aItem, b);
                }
            }
            return result;
        }
    }
    if (ac == null && bc == null) {
        // v/v
        if (vvf != null) {
            return vvf(a as VmConst[], b as VmConst[], ar, br);
        }
        const rr = max(ar, br);
        const result: VmConst[] = [];
        for (let r = 0; r < rr; r++) {
            const aItem = (a as VmConst[])[r] ?? null;
            const bItem = (b as VmConst[])[r] ?? null;
            result[r] = f(aItem, bItem);
        }
        return result;
    }

    // m/m (m/v v/m)
    if (ac == null) {
        // v/m
        if (vmf != null) {
            return vmf(a as VmConst[], b as VmConst[][], ar, br, bc!);
        }
        ac = ar!;
        ar = 1;
        a = [a];
    }
    if (bc == null) {
        // m/v
        if (mvf != null) {
            return mvf(a as VmConst[][], b as VmConst[], ar, ac, br);
        }
        bc = br!;
        br = 1;
        b = [b];
    }

    if (mmf != null) {
        return mmf(a as VmConst[][], b as VmConst[][], ar, ac, br, bc);
    }
    const rr = max(ar, br);
    const rc = max(ac, bc);
    const result: VmConst[][] = [];
    for (let r = 0; r < rr; r++) {
        const rRow: VmConst[] = [];
        result[r] = rRow;
        for (let c = 0; c < rc; c++) {
            const aItem = (a as VmConst[][])[ar === 1 ? 0 : r]?.[ac === 1 ? 0 : c] ?? null;
            const bItem = (b as VmConst[][])[br === 1 ? 0 : r]?.[bc === 1 ? 0 : c] ?? null;
            rRow[c] = f(aItem, bItem);
        }
    }
    return result;
}

export const entrywise = VmLib(
    (a, b, f) => {
        expectConst('a', a, null);
        expectConst('b', b, null);
        expectCallable('f', f, null);
        return entrywiseImpl(a, b, (a, b) => {
            Cp();
            const ret = $Call(f, [a, b]);
            if (!isVmConst(ret)) return null;
            return ret;
        });
    },
    {
        summary: '逐项操作',
        params: {
            a: { type: 'any | any[] | any[][]', description: '第一个操作数' },
            b: { type: 'any | any[] | any[][]', description: '第二个操作数' },
            f: { type: 'fn(a: any, b: any) -> any', description: '操作函数' },
        },
        returns: { type: 'any | any[] | any[][]' },
        examples: [`matrix.entrywise([1, 2], [3, 4], fn (x, y) { x + y }) // [4, 6]`],
    },
);

export const add = VmLib(
    (a, b) => {
        expectConst('a', a, null);
        expectConst('b', b, null);
        return entrywiseImpl(a, b, $Add);
    },
    {
        summary: '逐项相加',
        params: {
            a: { type: 'number | number[] | number[][]', description: '第一个操作数' },
            b: { type: 'number | number[] | number[][]', description: '第二个操作数' },
        },
        returns: { type: 'number | number[] | number[][]' },
        examples: ['matrix.add([1, 2], [3, 4]) // [4, 6]'],
    },
);

export const subtract = VmLib(
    (a, b) => {
        expectConst('a', a, null);
        expectConst('b', b, null);
        return entrywiseImpl(a, b, $Sub);
    },
    {
        summary: '逐项相减',
        params: {
            a: { type: 'number | number[] | number[][]', description: '第一个操作数' },
            b: { type: 'number | number[] | number[][]', description: '第二个操作数' },
        },
        returns: { type: 'number | number[] | number[][]' },
        examples: ['matrix.subtract([3, 4], [1, 2]) // [2, 2]'],
    },
);

export const entrywise_multiply = VmLib(
    (a, b) => {
        expectConst('a', a, null);
        expectConst('b', b, null);
        return entrywiseImpl(a, b, $Mul);
    },
    {
        summary: '逐项相乘',
        params: {
            a: { type: 'number | number[] | number[][]', description: '第一个操作数' },
            b: { type: 'number | number[] | number[][]', description: '第二个操作数' },
        },
        returns: { type: 'number | number[] | number[][]' },
        examples: ['matrix.entrywise_multiply([1, 2], [3, 4]) // [3, 8]'],
    },
);

export const entrywise_divide = VmLib(
    (a, b) => {
        expectConst('a', a, null);
        expectConst('b', b, null);
        return entrywiseImpl(a, b, $Div);
    },
    {
        summary: '逐项相除',
        params: {
            a: { type: 'number | number[] | number[][]', description: '第一个操作数' },
            b: { type: 'number | number[] | number[][]', description: '第二个操作数' },
        },
        returns: { type: 'number | number[] | number[][]' },
        examples: ['matrix.entrywise_divide([4, 6], [2, 3]) // [2, 2]'],
    },
);

export const multiply = VmLib(
    (a, b) => {
        expectConst('a', a, null);
        expectConst('b', b, null);
        return entrywiseImpl(
            a,
            b,
            $Mul,
            (a, b, al, bl) => {
                const l = max(al, bl);
                let s = 0;
                for (let i = 0; i < l; i++) {
                    s += num(a[i]) * num(b[i]);
                }
                return s;
            },
            (a, b, ar, ac, br, bc) => {
                if (ac !== br) throwError(`Incompatible matrix dimensions`, null);
                const result: VmConst[][] = [];
                for (let r = 0; r < ar; r++) {
                    const rRow: VmConst[] = [];
                    result[r] = rRow;
                    for (let c = 0; c < bc; c++) {
                        let item = 0;
                        for (let k = 0; k < ac; k++) {
                            item += num((a as VmConst[][])[r]?.[k]) * num((b as VmConst[][])[k]?.[c]);
                        }
                        rRow[c] = item;
                    }
                }
                return result;
            },
            (a, b, al, br, bc) => {
                if (al !== br) throwError(`Incompatible matrix dimensions`, null);
                const result: VmConst[] = [];
                for (let c = 0; c < bc; c++) {
                    let item = 0;
                    for (let k = 0; k < al; k++) {
                        item += num(a[k]) * num((b as VmConst[][])[k]?.[c]);
                    }
                    result[c] = item;
                }
                return result;
            },
            (a, b, ar, ac, bl) => {
                if (ac !== bl) throwError(`Incompatible matrix dimensions`, null);
                const result: VmConst[] = [];
                for (let r = 0; r < ar; r++) {
                    let item = 0;
                    for (let k = 0; k < ac; k++) {
                        item += num((a as VmConst[][])[r]?.[k]) * num(b[k]);
                    }
                    result[r] = item;
                }
                return result;
            },
        );
    },
    {
        summary: '矩阵相乘',
        params: {
            a: { type: 'number | number[] | number[][]', description: '第一个操作数' },
            b: { type: 'number | number[] | number[][]', description: '第二个操作数' },
        },
        returns: { type: 'number | number[] | number[][]' },
        examples: ['matrix.multiply([[1, 2], [3, 4]], [5, 6]) // [17, 39]'],
    },
);

File: crates/python/mirascript/_vm/lib/vm_global/mod/matrix/bin_ops.py
from __future__ import annotations

from ......_helpers.types import is_vm_const
from .....operations import Add, Call, Div, Mul, Sub, Cp
from ...._helpers import _expect_callable, _expect_const, _throw_error
from ._helper import _size, _num

__all__ = [
    "entrywise",
    "add",
    "subtract",
    "entrywise_multiply",
    "entrywise_divide",
    "multiply",
]


def _entrywise_impl(a, b, f, vvf=None, mmf=None, vmf=None, mvf=None):
    aDims = _size(a)
    bDims = _size(b)

    if len(aDims) == 0:
        if len(bDims) == 0:
            return f(a, b)
        elif len(bDims) == 1:
            bLen = bDims[0]
            result = []
            for j in range(bLen):
                bItem = b[j] if j < len(b) else None
                result.append(f(a, bItem))
            return result
        else:
            numRows, numCols = bDims
            result = []
            for i in range(numRows):
                Cp()
                newRow = []
                for j in range(numCols):
                    row = b[i] if i < len(b) else None
                    bItem = row[j] if row and j < len(row) else None
                    newRow.append(f(a, bItem))
                result.append(newRow)
            return result

    if len(bDims) == 0:
        if len(aDims) == 1:
            aLen = aDims[0]
            result = []
            for i in range(aLen):
                aItem = a[i] if i < len(a) else None
                result.append(f(aItem, b))
            return result
        else:
            numRows, numCols = aDims
            result = []
            for i in range(numRows):
                Cp()
                newRow = []
                for j in range(numCols):
                    row = a[i] if i < len(a) else None
                    aItem = row[j] if row and j < len(row) else None
                    newRow.append(f(aItem, b))
                result.append(newRow)
            return result
    if len(aDims) == 1 and len(bDims) == 1:
        if vvf is not None:
            return vvf(a, b, aDims[0], bDims[0])
        aLen = aDims[0]
        bLen = bDims[0]
        result = []
        rr = max(aLen, bLen)
        for i in range(rr):
            aItem = a[i] if i < len(a) else None
            bItem = b[i] if i < len(b) else None
            result.append(f(aItem, bItem))
        return result

    if len(aDims) == 1:
        if vmf is not None:
            return vmf(a, b, aDims[0], bDims[0], bDims[1])
        aLen = aDims[0]
        numRows, numCols = bDims
        if aLen != numCols:
            _throw_error("Array length mismatch", [])
        result = []
        for i in range(numRows):
            newRow = []
            for j in range(numCols):
                aItem = a[j] if j < len(a) else None
                row = b[i] if i < len(b) else None
                bItem = row[j] if row and j < len(row) else None
                newRow.append(f(aItem, bItem))
            result.append(newRow)
        return result
    if len(bDims) == 1:
        if mvf is not None:
            return mvf(a, b, aDims[0], aDims[1], bDims[0])
        bLen = bDims[0]
        numRows, numCols = aDims
        if bLen != numCols:
            _throw_error("Array length mismatch", [])
        result = []
        for i in range(numRows):
            newRow = []
            for j in range(numCols):
                row = a[i] if i < len(a) else None
                aItem = row[j] if row and j < len(row) else None
                bItem = b[j] if j < len(b) else None
                newRow.append(f(aItem, bItem))
            result.append(newRow)
        return result
    if mmf is not None:
        return mmf(a, b, aDims[0], aDims[1], bDims[0], bDims[1])

    rr = max(aDims[0], bDims[0])
    cc = max(aDims[1], bDims[1])
    result = []
    for i in range(rr):
        newRow = []
        for j in range(cc):
            ar = 0 if aDims[0] == 1 else i
            ac = 0 if aDims[1] == 1 else j
            br = 0 if bDims[0] == 1 else i
            bc = 0 if bDims[1] == 1 else j
            aRow = a[ar] if ar < len(a) else None
            aItem = aRow[ac] if aRow and ac < len(aRow) else None
            bRow = b[br] if br < len(b) else None
            bItem = bRow[bc] if bRow and bc < len(bRow) else None
            newRow.append(f(aItem, bItem))

            # aRow = a[i] if  i < len(a) else None
            # aItem = aRow[j] if aRow and  j < len(aRow) else None
            # bRow = b[i] if i < len(b) else None
            # bItem = bRow[j] if bRow and  j < len(bRow) else None
            # newRow.append(f(aItem,bItem))
        result.append(newRow)
    return result


def entrywise(matrix, scalar, fn):
    _expect_const("matrix", matrix, [])
    _expect_const("scalar", scalar, [])
    _expect_callable("fn", fn, [])

    def f(a, b):
        ret = Call(fn, a, b)
        if not is_vm_const(ret):
            return None
        return ret

    return _entrywise_impl(matrix, scalar, f)


def add(a, b):
    _expect_const("a", a, [])
    _expect_const("b", b, [])
    return _entrywise_impl(a, b, Add)


def subtract(a, b):
    _expect_const("a", a, [])
    _expect_const("b", b, [])
    return _entrywise_impl(a, b, Sub)


def entrywise_multiply(a, b):
    _expect_const("a", a, [])
    _expect_const("b", b, [])
    return _entrywise_impl(a, b, Mul)


def entrywise_divide(a, b):
    _expect_const("a", a, [])
    _expect_const("b", b, [])
    return _entrywise_impl(a, b, Div)


def multiply(a, b):
    _expect_const("a", a, [])
    _expect_const("b", b, [])

    def vvf(a, b, aLen, bLen):
        rr = max(aLen, bLen)
        s = 0
        for i in range(rr):
            aItem = a[i] if i < len(a) else None
            bItem = b[i] if i < len(b) else None
            s += _num(aItem) * _num(bItem)

        return s

    def mmf(a, b, aRows, aCols, bRows, bCols):
        if aCols != bRows:
            _throw_error("Matrix size mismatch for multiplication", [])
        result = []
        for i in range(aRows):
            newRow = []
            for j in range(bCols):
                sum = 0
                for k in range(aCols):
                    aRow = a[i] if i < len(a) else None
                    aItem = aRow[k] if aRow and k < len(aRow) else None
                    bRow = b[k] if k < len(b) else None
                    bItem = bRow[j] if bRow and j < len(bRow) else None
                    sum = Add(sum, Mul(aItem, bItem))
                newRow.append(sum)
            result.append(newRow)
        return result

    def vmf(a, b, aLen, bRows, bCols):
        if aLen != bRows:
            _throw_error("Vector and matrix size mismatch for multiplication", [])

        result = []
        for i in range(bCols):
            item = 0
            for j in range(aLen):
                aItem = a[j] if j < len(a) else None
                bRow = b[j] if j < len(b) else None
                bItem = bRow[i] if bRow and i < len(bRow) else None
                # newRow.append(Mul_(aItem, bItem))
                item += _num(aItem) * _num(bItem)
            result.append(item)
        return result

    def mvf(a, b, aRows, aCols, bLen):
        if aCols != bLen:
            _throw_error("Matrix and vector size mismatch for multiplication", [])
        result = []
        for i in range(aRows):
            sum = 0
            for j in range(aCols):
                aRow = a[i] if i < len(a) else None
                aItem = aRow[j] if aRow and j < len(aRow) else None
                bItem = b[j] if j < len(b) else None
                sum = Add(sum, Mul(aItem, bItem))
            result.append(sum)
        return result

    return _entrywise_impl(a, b, Mul, vvf=vvf, mmf=mmf, vmf=vmf, mvf=mvf)
